base/src/xml.ext.c

```diff
@@ -12,6 +12,92 @@
     (nm)->str = acton_malloc_atomic((nm)->nbytes + 1);       \
     (nm)->str[(nm)->nbytes] = 0
 
+// Helper function to count extra bytes needed for XML escaping
+// Returns the number of extra bytes needed to replace the special character
+// with encoded entity (not total bytes)
+static int count_xml_escape_extra(unsigned char *str, int nbytes, int escape_quotes) {
+    int extra = 0;
+    for (int i = 0; i < nbytes; i++) {
+        switch (str[i]) {
```

Comment:
I think this is wrong. You are iterating over this byte-by-byte. XML takes strings as input which allow utf-8 chars, so you could be inspecting the second byte (or third or four) in a utf-8 char and think that it's a special char when it's really not.

The comparison is basically correct, I just think you need to iterate over (utf-8) chars. Look at `utf8proc_iterate` - it's used in `str.c`. You could also consider writing such escape functions in Acton instead of in C.